### PR TITLE
refactor iterator

### PR DESCRIPTION
**Pull request by BusyJay**
@ngaut @siddontang PTAL

### CODE REVIEW

File: src/rocksdb.rs [hunk 1]
```diff
@@ -54,69 +54,37 @@ pub struct DBIterator<'a> {
...
+        let kv = self.kv();
+        if kv.is_some() {
+            DBIterator::next(self);
         }
+        kv
     }
 }
 
-pub enum IteratorMode<'a> {
+pub enum Anchor<'a> {
```
Comment (siddontang) [inline src/rocksdb.rs]: I don't think Anchor is better here. 
I guess this is more similar like Whence in C seek.

**Comment by BusyJay** [comment]
@disksing PTAL

File: src/rocksdb.rs [hunk 2]
```diff
@@ -53,108 +53,98 @@ pub struct Snapshot<'a> {
...
+        let val_len_ptr: *mut size_t = &mut val_len;
+        unsafe {
+            let val_ptr = rocksdb_ffi::rocksdb_iter_value(self.inner,
+                                                          val_len_ptr);
+            slice::from_raw_parts(val_ptr, val_len as usize)
+        }
+    }
+
+    pub fn kv(&self) -> Option<(Vec<u8>, Vec<u8>)> {
+        if self.valid() {
```
Comment (disksing) [inline src/rocksdb.rs]: Why check `valid()` for `kv()` but not for `key()` and `value()`?

File: src/rocksdb.rs [hunk 3]
```diff
@@ -53,108 +53,98 @@ pub struct Snapshot<'a> {
...
+        } else {
+            None
+        }
+    }
+
+    pub fn prev(&mut self) -> bool {
+        unsafe {
+            rocksdb_ffi::rocksdb_iter_prev(self.inner);
+        }
+        self.valid()
```
Comment (disksing) [inline src/rocksdb.rs]: Better to put `prev()` next to `next()`.

File: src/rocksdb.rs [hunk 4]
```diff
@@ -175,17 +165,29 @@ impl<'a> DBIterator<'a> {
...
+            rv.seek(key);
 
             Ok(rv)
         }
     }
 }
 
+pub type Kv = (Vec<u8>, Vec<u8>);
+
+impl<'a> Iterator for DBIterator<'a> {
```
Comment (disksing) [inline src/rocksdb.rs]: We may impl `DoubleEndedIterator ` which can seek back.

File: src/rocksdb.rs (hunk 2 above)
Comment (BusyJay) [inline src/rocksdb.rs]: Actually `key()` and `value()` also check `valid()` too. But `key()` and `value()` require `valid()` is true.

It's convenient to omit Option, and `valid()` is often checked before calling `key()` or `value()`.

File: src/rocksdb.rs (hunk 4 above)
Comment (BusyJay) [inline src/rocksdb.rs]: No, `DoubleEndedIterator` requires to keep track both end's seek point, which is inefficient.

Comment (disksing) [inline src/rocksdb.rs]: Got it. I misunderstood it.

**Comment by disksing** [comment]
LGTM

**Comment by ngaut** [comment]
LGTM

**Comment by siddontang** [comment]
LGTM
I think this fork is very different from origin one now :-)

**Pull request merged**